src/viper/trans.ml

```diff
@@ -312,6 +337,22 @@ and stmt ctxt (s : M.exp) : seqn =
        accE at (self ctxt Source.no_region, id)
      in
      ctxt.ghost_perms := mk_p :: !(ctxt.ghost_perms);
+     let stmts = stmt ctxt e in
+     (* assume that each `async {...}` has an assertion *)
+     let conc, _ = extract_concurrency stmts in
+     let mk_c = match conc with
+       | [] ->
+         fun _ x -> x
+       | ConcurrencyS ("1", _, cond) :: _ ->
+         let (!!) p = !!! (cond.at) p in
+         let zero, one = intLitE Source.no_region 0, intLitE Source.no_region 1 in
+         fun ctxt x ->
+           let ghost_fld = !!(FldAcc (self ctxt Source.no_region, id)) in
+           let between = !!(AndE (!!(LeCmpE (zero, ghost_fld)), !!(LeCmpE (ghost_fld, one)))) in
+           let is_one = !!(EqCmpE (ghost_fld, one)) in
```

Comment:
```suggestion
           let ghost_fld () = !!(FldAcc (self ctxt Source.no_region, id)) in
           let between = !!(AndE (!!(LeCmpE (zero, ghost_fld ())), !!(LeCmpE (ghost_fld (), one)))) in
           let is_one = !!(EqCmpE (ghost_fld (), one)) in
```
To avoid sharing mutable notes (which might be problem in the future, but not now)

Reply:
Me getting lazy because of Haskell exposure... Yep, makes sense!